### PR TITLE
INT-4236: Status for SmartLifecycleRoleControlller

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-4236

Add methods to obtain status from the `SmartLifecycleRoleController`.

__backport will require removal of diamonds and stream operations - I will issue a separate PR after review/merge__

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java [hunk 1]
```diff
@@ -185,7 +190,76 @@ public void stopLifecyclesInRole(String role) {
...
+	 * @since 4.3.8
+	 */
+	public boolean allEndpointsRunning(String role) {
+		Map<String, Boolean> status = getEndpointsRunningStatus(role);
+		if (status.isEmpty()) {
+			return false;
+		}
+		else {
+			return status.values().stream().allMatch(b -> b);
+		}
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java]: Can be simplified:
```
return !status.isEmpty() && status.values().stream().allMatch(b -> b);
```

File: spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java [hunk 2]
```diff
@@ -185,7 +190,76 @@ public void stopLifecyclesInRole(String role) {
...
+	 * @return true if there are no endpoints or none are running.
+	 * @since 4.3.8
+	 */
+	public boolean noEndpointsRunning(String role) {
+		Map<String, Boolean> status = getEndpointsRunningStatus(role);
+		if (status.isEmpty()) {
+			return true;
+		}
+		else {
+			return status.values().stream().allMatch(b -> !b);
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java]: Negative ``allMatch` can be replaced with the `noneMatch`.
Plus `if...else` can be simplified too:
```
return status.isEmpty() || status.values().stream().noneMatch(b -> b);
```
That is what IDEA does for me! 😄

File: spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java [hunk 3]
```diff
@@ -185,7 +190,76 @@ public void stopLifecyclesInRole(String role) {
...
+			addLazyLifecycles();
+		}
+		if (!this.lifecycles.containsKey(role)) {
+			return Collections.emptyMap();
+		}
+		Map<String, Boolean> runningStatus = new HashMap<>();
+		AtomicInteger index = new AtomicInteger();
+		this.lifecycles.get(role)
+				.stream()
+				.forEach(e -> {
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java]: Just in case:
redundant_stream

Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java]: OTOH, there is something like this:
````java
return this.lifecycles.get(role)
		.stream()
		.collect(Collectors.toMap(e ->
						(e instanceof NamedComponent)
								? ((NamedComponent) e).getComponentName()
								: (e.getClass().getSimpleName() + "#" + index.getAndIncrement()),
				SmartLifecycle::isRunning));
````
http://stackoverflow.com/questions/20363719/java-8-listv-into-mapk-v

File: src/reference/asciidoc/endpoint.adoc [hunk 4]
```diff
@@ -633,6 +633,27 @@ start/stop its configured `SmartLifecycle` objects when leadership is granted/re
...
+public Map<String, Boolean> getEndpointsRunningStatus(String role) <4>
+----
+
+<1> Returns a list of the roles being managed.
+
+<2> Returns true if all endpoints in the role are running.
+
+<3> Returns true if none of the endpoints in the role are running.
+
+<4> Returns a map of `compoonent name : running status` - the component name is usually the bean name.
```
Comment (artembilan) [inline src/reference/asciidoc/endpoint.adoc]: Typo: comp~~o~~onent

**Review by artembilan — CHANGES_REQUESTED** [review]
That's all
I have all those changes locally.
Can merge if you agree.

**Comment by garyrussell** [comment]
Good; thanks!

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/f23360d53955e2e1ca13c8f462680e5562cd2a1a.
Pay attention how I have fixed log message. Might be useful for backport PR.

**Pull request closed without merging**